cpp/oneapi/dal/algo/svm/backend/gpu/working_set_selector.hpp

```diff
@@ -0,0 +1,266 @@
+                              const pr::ndview<Float, 1>& y,
+                              const pr::ndview<Float, 1>& alpha,
+                              pr::ndview<std::uint8_t, 1>& indicator,
+                              const Float C,
+                              const std::int64_t n,
+                              ws_edge edge,
+                              const dal::backend::event_vector& deps = {}) {
+        ONEDAL_ASSERT(y.get_dimension(0) == n);
+        ONEDAL_ASSERT(alpha.get_dimension(0) == n);
+        ONEDAL_ASSERT(indicator.get_dimension(0) == n);
+        ONEDAL_ASSERT(indicator.has_mutable_data());
+
+        const Float* y_ptr = y.get_data();
+        const Float* alpha_ptr = alpha.get_data();
+        std::uint8_t* indicator_ptr = indicator.get_mutable_data();
+
+        const auto wg_size = std::min(dal::backend::propose_wg_size(queue), n);
+        const auto range = dal::backend::make_multiple_nd_range_1d(n, wg_size);
+
+        sycl::event check_event;
+
+        if (edge == ws_edge::up) {
+            check_event = queue.submit([&](sycl::handler& cgh) {
+                cgh.depends_on(deps);
+
+                cgh.parallel_for(range, [=](sycl::nd_item<1> item) {
+                    const std::uint32_t i = item.get_global_id(0);
+                    indicator_ptr[i] = is_upper_edge<Float>(y_ptr[i], alpha_ptr[i], C);
+                });
+            });
+        }
+        else {
+            check_event = queue.submit([&](sycl::handler& cgh) {
+                cgh.depends_on(deps);
+
+                cgh.parallel_for(range, [=](sycl::nd_item<1> item) {
+                    const std::uint32_t i = item.get_global_id(0);
+                    indicator_ptr[i] = is_lower_edge<Float>(y_ptr[i], alpha_ptr[i], C);
+                });
+            });
+        }
+        return check_event;
+    }
+
+    std::tuple<sycl::event, const std::int64_t> copy_last_to_first(
+        pr::ndview<std::uint32_t, 1>& ws_indices,
+        const dal::backend::event_vector& deps = {}) {
+        ONEDAL_ASSERT(ws_indices.get_dimension(0) == ws_count_);
+        ONEDAL_ASSERT(ws_indices.has_mutable_data());
+        const std::int64_t q = ws_count_ / 2;
+        std::uint32_t* ws_indices_ptr = ws_indices.get_mutable_data();
+        auto copy_event =
+            dal::backend::copy(queue_, ws_indices_ptr, ws_indices_ptr + q, ws_count_ - q, deps);
+        const std::int64_t selected_count = q;
+        return { copy_event, selected_count };
+    }
+
+    sycl::queue queue_;
+
+    std::int64_t row_count_;
+    std::int64_t ws_count_;
+    Float C_;
+
+    pr::ndarray<std::uint32_t, 1> sorted_f_indices_;
+    pr::ndarray<std::uint32_t, 1> buff_indices_;
+    pr::ndarray<std::uint8_t, 1> indicator_;
+    pr::ndarray<Float, 1> values_sort_;
+    pr::ndarray<Float, 1> labels_;
+};
+
+#define INSTANTIATE_WORKING_SET(F) template class working_set_selector<F>;
+
+INSTANTIATE_WORKING_SET(float);
+INSTANTIATE_WORKING_SET(double);
```

Comment:
I think this need to be moved to cpp file. Can you also move method implementations to the cpp?